src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -643,55 +626,155 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                arrayType = arrayType.WithElementType(TypeWithAnnotations.Create(elementType.EnumUnderlyingTypeOrSelf()));
+
+                var cachingField = _builder.module.GetArrayCachingFieldForData(data, _module.Translate(arrayType), wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                var arrayNotNullLabel = new object();
+
+                // T[]? array = PrivateImplementationDetails.cachingField;
+                // if (array is not null) goto arrayNotNull;
+                _builder.EmitOpCode(ILOpCode.Ldsfld);
+                _builder.EmitToken(cachingField, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Dup);
+                _builder.EmitBranch(ILOpCode.Brtrue, arrayNotNullLabel);
+
+                // array = new T[elementCount];
+                // RuntimeHelpers.InitializeArray(token, array);
+                // PrivateImplementationDetails.cachingField = array;
+                _builder.EmitOpCode(ILOpCode.Pop);
+                _builder.EmitIntConstant(elementCount);
+                _builder.EmitOpCode(ILOpCode.Newarr);
+                EmitSymbolToken(arrayType.ElementType, wrappedExpression.Syntax);
+                _builder.EmitArrayBlockInitializer(data, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Dup);
+                _builder.EmitOpCode(ILOpCode.Stsfld);
+                _builder.EmitToken(cachingField, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+
+                // arrayNotNullLabel:
+                // new ReadOnlySpan<T>(array)
+                _builder.MarkLabel(arrayNotNullLabel);
+                _builder.EmitOpCode(ILOpCode.Newobj, 0);
+                EmitSymbolToken(rosArrayCtor.AsMember(spanType), wrappedExpression.Syntax, optArgList: null);
+                return true;
+            }
+
+            // Emit: new ReadOnlySpan<ElementType>(PrivateImplementationDetails.cachingField ??= new ElementType[] { ... constants ... })
+            bool tryEmitAsCachedArrayOfConstants(BoundArrayCreation arrayCreation, ArrayTypeSymbol arrayType, TypeSymbol elementType, NamedTypeSymbol spanType, bool used, BoundExpression? inPlaceTarget, out bool avoidInPlace)
+            {
+                avoidInPlace = false;
+                var initializer = arrayCreation.InitializerOpt;
+                Debug.Assert(initializer != null);
+
+                var initializers = initializer.Initializers;
+                Debug.Assert(initializers.All(static init => init.ConstantValueOpt != null));
+                Debug.Assert(!elementType.IsEnumType());
+
+                if (!tryGetReadOnlySpanArrayCtor(arrayCreation.Syntax, out var rosArrayCtor))
+                {
+                    return false;
+                }
+
+                if (inPlaceTarget is not null)
+                {
+                    EmitAddress(inPlaceTarget, Binder.AddressKind.Writeable);
```

Comment:
>EmitAddress [](http://example.com/codeflow?start=20&length=11)

Just curious: Is this code path hit for collection expressions only?

Reply:
No, we can also reach it with code like this (I'll add to a test):
```
        System.ReadOnlySpan<object> s4;
        Print(s4 = new System.ReadOnlySpan<object>(new object[] {{ null, null }}));
```